### PR TITLE
Template-based decorators for PyBind layer

### PR DESCRIPTION
**Pull request by billschereriii**
WIP, showing direction

### CODE REVIEW

**Comment by billschereriii** [comment]
Transformation complete

File: src/python/src/pyclient.cpp [hunk 1]
```diff
@@ -52,424 +56,233 @@ PyClient::PyClient(bool cluster, const std::string& logger_name)
...
+    });
 }
 
-void PyClient::rename_dataset(const std::string& old_name,
-                             const std::string& new_name) {
-    try {
+void PyClient::rename_dataset(
+    const std::string& old_name, const std::string& new_name)
+{
+    MAKE_CLIENT_API("copy_dataset", {
```
[the remark this answers is not in the document]
Comment (MattToast) [reply src/python/src/pyclient.cpp:227]: per convo, we might be able to do something with `__func__`

**Review by MattToast — CHANGES_REQUESTED** [review]
Looks amazing!! So much boilerplate removed! Caught a couple of misnamed `MAKE_CLIENT_API` calls, but overall everything here looks fantastic!

File: src/python/src/pyclient.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/src/pyclient.cpp]: done and done

File: src/python/src/pyclient.cpp [hunk 2]
```diff
@@ -480,86 +293,33 @@ void PyClient::run_script_multigpu(const std::string& name,
...
-    catch (std::exception& e) {
-        // should never happen
-        throw SRInternalException(e.what());
-    }
-    catch (...) {
-        // should never happen
-        throw SRInternalException("A non-standard exception was encountered "\
-                                  "while executing run_script_multigpu.");
-    }
+    MAKE_CLIENT_API("run_script", {
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/src/pyclient.cpp]: moot with use of __func__

**Review by MattToast — APPROVED** [review]
This looks amazing!! Thanks for all the hard work on this, this removes soooo much repetitive boilerplate!! LGTM pending tests!!

**Pull request merged**